pkg/collect/velero.go

```diff
@@ -0,0 +1,479 @@
+package collect
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/http"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+	troubleshootv1beta2 "github.com/replicatedhq/troubleshoot/pkg/apis/troubleshoot/v1beta2"
+	veleroclient "github.com/vmware-tanzu/velero/pkg/generated/clientset/versioned"
+
+	"gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	apiErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+)
+
+const (
+	DefaultVeleroNamespace = "velero"
+)
+
+type CollectVelero struct {
+	Collector    *troubleshootv1beta2.Velero
+	BundlePath   string
+	Namespace    string
+	ClientConfig *rest.Config
+	Client       kubernetes.Interface
+	Context      context.Context
+	RBACErrors
+}
+
+type VeleroCommand struct {
+	ID             string
+	Command        []string
+	Args           []string
+	Format         string
+	DefaultTimeout string
+}
+
+var VeleroCommands = []VeleroCommand{
+	{
+		ID:             "get-backups",
+		Command:        []string{"/velero", "get", "backups"},
+		Args:           []string{"-o", "yaml"},
+		Format:         "yaml",
+		DefaultTimeout: "30s",
+	},
+	{
+		ID:             "get-restores",
+		Command:        []string{"/velero", "get", "restores"},
+		Args:           []string{"-o", "yaml"},
+		Format:         "yaml",
+		DefaultTimeout: "30s",
+	},
+	{
+		ID:             "describe-backups",
+		Command:        []string{"/velero", "describe", "backups"},
+		Args:           []string{"--details", "--colorized", "false"},
+		Format:         "txt",
+		DefaultTimeout: "30s",
+	},
+	{
+		ID:             "describe-restores",
+		Command:        []string{"/velero", "describe", "restores"},
+		Args:           []string{"--details", "--colorized", "false"},
+		Format:         "txt",
+		DefaultTimeout: "30s",
+	},
+}
+
+func (c *CollectVelero) Title() string {
+	return getCollectorName(c)
+}
+
+func (c *CollectVelero) IsExcluded() (bool, error) {
+	return isExcluded(c.Collector.Exclude)
+}
+
+func (c *CollectVelero) Collect(progressChan chan<- interface{}) (CollectorResult, error) {
+	ctx := context.TODO()
+
+	ns := DefaultVeleroNamespace
+	if c.Collector.Namespace != "" {
+		ns = c.Collector.Namespace
+	}
+
+	pod, err := findVeleroPod(ctx, c, c.Collector.Namespace)
+	if err != nil {
+		return nil, err
+	}
+	output := NewResult()
+
+	// collect output from velero binary
+	if pod != nil {
+		for _, command := range VeleroCommands {
```

Comment:
do we need some logic to check that the velero binary exists on a system? eg. if the collector is defined but velero isn't installed, will this error out on bundle creation? (not ideal)

we should gracefully skip collection if the velero binary isn't found, and/or provide something informative. bonus points if its visible in the bundle that the binary wasn't found